File: scala/node/src/main/scala/com/giraffechain/p2p/SharedSync.scala
package com.giraffechain.p2p

import cats.data.OptionT
import cats.effect.implicits.*
import cats.effect.std.{Mutex, Random}
import cats.effect.{Async, Fiber, Ref, Resource}
import cats.implicits.*
import com.giraffechain.codecs.given
import com.giraffechain.consensus.ChainSelectionOutcome
import com.giraffechain.models.*
import com.giraffechain.{BlockchainCore, Height}
import fs2.{Pipe, Pull, Stream}
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger

class SharedSync[F[_]: Async: Random](
    core: BlockchainCore[F],
    clientsF: F[Map[PeerId, PeerBlockchainInterface[F]]],
    targetRef: Ref[F, Option[SharedSyncState[F]]],
    mutex: Mutex[F]
):

  private given logger: Logger[F] =
    Slf4jLogger.getLoggerFromName("SharedSync")

  def compareAndSync(commonAncestorHeader: BlockHeader, target: BlockHeader, peerId: PeerId): F[Unit] =
    OptionT(targetRef.get).foldF(
      localCompareAndSync(commonAncestorHeader, target, peerId)
    )(state =>
      if (state.target.id == target.id || state.target.id == target.parentHeaderId)
        updateTarget(commonAncestorHeader, target, peerId).void // .flatMap(_.fiber.joinWithUnit)
      else remoteCompareAndSync(commonAncestorHeader, target, peerId)(state.target, state.providers)
    )

  def omitPeer(peerId: PeerId): F[Unit] =
    targetRef.modify {
      case Some(state) if state.providers == Set(peerId) =>
        (none, state.fiber.cancel)
      case Some(state) if state.providers.contains(peerId) =>
        (state.copy(providers = state.providers - peerId).some, ().pure[F])
      case s =>
        (s, ().pure[F])
    }

  private def localCompareAndSync(commonAncestorHeader: BlockHeader, target: BlockHeader, peerId: PeerId) =
    mutex.lock
      .surround(
        for {
          interface <- clientsF.map(_.apply(peerId))
          remoteHeaderAtHeightF = (height: Height) =>
            OptionT(interface.blockIdAtHeight(height)).flatMapF(interface.fetchHeader).value
          localHeadId <- core.consensus.localChain.currentHead
          localHeader <- core.dataStores.headers.getOrRaise(localHeadId)
          localHeaderAtHeightF = (height: Height) =>
            OptionT(core.consensus.localChain.blockIdAtHeight(height)).flatMapF(core.dataStores.headers.get).value
          chainSelectionResult <- core.consensus.chainSelection.compare(
            localHeader,
            target,
            commonAncestorHeader,
            localHeaderAtHeightF,
            remoteHeaderAtHeightF
          )
          _ <- logResult(chainSelectionResult)
          newStateOpt <- OptionT
            .whenF(chainSelectionResult.isY)(updateTarget(commonAncestorHeader, target, peerId))
            .value
        } yield newStateOpt
      )
//      .flatTap(_.traverse(_.fiber.joinWithUnit))
      .void

  private def remoteCompareAndSync(commonAncestorHeader: BlockHeader, target: BlockHeader, peerId: PeerId)(
      currentTarget: BlockHeader,
      providers: Set[PeerId]
  ) =
    mutex.lock
      .surround(
        for {
          clients <- clientsF
          interface = clients(peerId)
          remoteHeaderAtHeightF = (height: Height) =>
            OptionT(interface.blockIdAtHeight(height)).flatMapF(interface.fetchHeader).value
          localInterface = new MultiPeerInterface[F](providers.map(clients).toList)
          localHeaderAtHeightF = (height: Height) =>
            OptionT(localInterface.blockIdAtHeight(height)).flatMapF(localInterface.fetchHeader).value
          chainSelectionResult <- core.consensus.chainSelection.compare(
            currentTarget,
            target,
            commonAncestorHeader,
            localHeaderAtHeightF,
            remoteHeaderAtHeightF
          )
          _ <- logResult(chainSelectionResult)
          newStateOpt <- OptionT
            .whenF(chainSelectionResult.isY)(updateTarget(commonAncestorHeader, target, peerId))
            .value
        } yield newStateOpt
      )
//      .flatTap(_.traverse(_.fiber.joinWithUnit))
      .void

  private def logResult(chainSelectionResult: ChainSelectionOutcome) =
    chainSelectionResult match {
      case ChainSelectionOutcome.XStandard =>
        Logger[F].info("Remote peer is up-to-date but local chain is better")
      case ChainSelectionOutcome.YStandard =>
        Logger[F].info("Local peer is up-to-date but remote chain is better")
      case ChainSelectionOutcome.XDensity =>
        Logger[F].info("Remote peer is out-of-sync but local chain is better")
      case ChainSelectionOutcome.YDensity =>
        Logger[F].info("Local peer out-of-sync and remote chain is better")
    }

  private def updateTarget(commonAncestor: BlockHeader, target: BlockHeader, provider: PeerId): F[SharedSyncState[F]] =
    targetRef.get
      .flatMap {
        case Some(s @ SharedSyncState(current, providers, _)) if current.id == target.id =>
          s.copy(providers = providers + provider).pure[F]
        case Some(s @ SharedSyncState(current, _, _)) if current.id == target.parentHeaderId =>
          s.copy(target = target, providers = Set(provider)).pure[F]
        case _ =>
          sync(commonAncestor).start
            .map(fiber => SharedSyncState(target, Set(provider), fiber))
      }
      .flatTap(state => targetRef.set(state.some))

  private def sync(commonAncestor: BlockHeader): F[Unit] =
    Stream
      .unfoldEval(commonAncestor.height + 1)(h =>
        OptionT(targetRef.get)
          .map(_.target)
          .filter(_.height >= h)
          .as((h, h + 1))
          .value
      )
      .chunkLimit(2048)
      .evalMap(heights =>
        for {
          clients <- clientsF
          lastHeight = heights.last.get
          SharedSyncState(_, providers, _) <- OptionT(targetRef.get).getOrRaise(
            new IllegalStateException("Target not set")
          )
          providerInterfaces = providers.map(clients).toList
          providersInterface = new MultiPeerInterface[F](providerInterfaces)
          batchTargetId <- OptionT(providersInterface.blockIdAtHeight(lastHeight)).getOrRaise(
            new IllegalStateException("Target not found")
          )
          alternativeClients <- (clients -- providers).values.toList.traverseFilter(client =>
            OptionT(client.blockIdAtHeight(lastHeight)).filter(_ == batchTargetId).as(client).value
          )
          interface = new MultiPeerInterface[F](providerInterfaces ++ alternativeClients)
        } yield (heights, interface)
      )
      .flatMap((heights, interface) =>
        Stream
          .chunk(heights)
          .parEvalMap(256)(height =>
            OptionT(interface.blockIdAtHeight(height)).getOrRaise(
              new IllegalStateException("Block at Height not found")
            )
          )
          .parEvalMap(256)(id =>
            OptionT(interface.fetchHeader(id))
              .getOrRaise(new IllegalArgumentException("Remote header not found"))
          )
          .through(noForks)
          .through(fetchVerifyPersistPipe(interface))
      )
      .through(adoptSparsely)
      .compile
      .drain

  private def fetchVerifyPersistPipe(interface: PeerBlockchainInterface[F]): Pipe[F, BlockHeader, FullBlock] =
    _.evalTap(PeerBlockchainHandler.checkHeader(core))
      .parEvalMap(256)(PeerBlockchainHandler.fetchFullBlock(core, interface))
      .evalTap(PeerBlockchainHandler.checkBody(core))

  private def noForks: Pipe[F, BlockHeader, BlockHeader] = {
    def start(s: Stream[F, BlockHeader]): Pull[F, BlockHeader, Unit] =
      s.pull.uncons1.flatMap {
        case Some((head, tail)) =>
          Pull.output1(head) >> go(tail, head)
        case None =>
          Pull.done
      }

    def go(
        s: Stream[F, BlockHeader],
        previous: BlockHeader
    ): Pull[F, BlockHeader, Unit] =
      s.pull.uncons1.flatMap {
        case Some((head, tlStream)) =>
          if (head.parentHeaderId != previous.id)
            Pull.raiseError(new IllegalStateException("Remote peer branched during sync"))
          else Pull.output1(head) >> go(tlStream, head)
        case None =>
          Pull.done
      }

    start(_).stream
  }

  private def adoptSparsely: Pipe[F, FullBlock, FullBlock] = {
    val epochThirdLength = core.clock.epochLength / 3L
    def start(s: Stream[F, FullBlock]): Pull[F, FullBlock, Unit] =
      s.pull.uncons1.flatMap {
        case Some((head, tail)) =>
          Pull.output1(head) >> go(tail, head, false, head.header.slot / epochThirdLength)
        case None =>
          Pull.done
      }

    def go(
        s: Stream[F, FullBlock],
        previous: FullBlock,
        previousWasAdopted: Boolean,
        lastAdoptedThird: Long
    ): Pull[F, FullBlock, Unit] =
      s.pull.uncons1.flatMap {
        case Some((head, tail)) =>
          val third = head.header.slot / epochThirdLength
          if (third != lastAdoptedThird)
            Pull.eval(PeerBlockchainHandler.adoptAndLog(core)(head)) >> Pull
              .output1(head) >> go(tail, head, true, third)
          else
            Pull.output1(head) >> go(tail, head, false, lastAdoptedThird)
        case None =>
          if (previousWasAdopted) Pull.done
          else Pull.eval(PeerBlockchainHandler.adoptAndLog(core)(previous)) >> Pull.done
      }

    start(_).stream
  }

object SharedSync:
  def make[F[_]: Async: Random](
      core: BlockchainCore[F],
      clientsF: F[Map[PeerId, PeerBlockchainInterface[F]]]
  ): Resource[F, SharedSync[F]] =
    (
      Resource.make(Ref.of[F, Option[SharedSyncState[F]]](None))(
        _.getAndSet(none).flatMap(_.traverse(_.fiber.cancel)).void
      ),
      Mutex[F].toResource
    )
      .mapN(new SharedSync(core, clientsF, _, _))

case class SharedSyncState[F[_]](target: BlockHeader, providers: Set[PeerId], fiber: Fiber[F, Throwable, Unit])

File: dart/protocol/lib/src/blockchain/p2p/shared_sync.dart
import 'dart:async';
import 'dart:collection';

import 'package:fixnum/fixnum.dart';
import 'package:giraffe_protocol/protocol.dart';
import 'package:giraffe_sdk/sdk.dart';
import 'package:logging/logging.dart';
import 'package:mutex/mutex.dart';

import 'peer_handler.dart';

class SharedSync {
  final BlockchainCore core;
  final Map<PeerId, PeerBlockchainInterface> Function() clientsF;
  SharedSyncState? _state;
  final mutex = Mutex();

  SharedSync({required this.core, required this.clientsF});

  Future<void> compare(
          BlockHeader commonAncestor, BlockHeader target, PeerId peerId) =>
      mutex.protect(() async {
        if (_state == null) {
          return localCompare(commonAncestor, target, peerId);
        } else if (_state!.target.id == target.id ||
            _state!.target.id == target.parentHeaderId) {
          return updateTarget(commonAncestor, target, peerId);
        } else {
          return remoteCompare(target, peerId, _state!);
        }
      });

  Future<void> omitPeer(PeerId peerId) async {
    if (_state != null) {
      if (_state!.providers.length == 1 && _state!.providers.contains(peerId)) {
        await _state!.cancel();
        _state = null;
      } else {
        _state!.providers.remove(peerId);
      }
    }
  }

  Future<void> syncCompletion() async {
    if (_state != null) {
      await _state!.result;
      _state = null;
    }
  }

  Future<void> localCompare(
      BlockHeader commonAncestor, BlockHeader target, PeerId peerId) async {
    final clients = clientsF();
    final interface = clients[peerId];
    if (interface == null) {
      throw StateError("Peer not found");
    }
    final localHeader = await core.dataStores.headers
        .getOrRaise(core.consensus.localChain.head);

    Future<BlockHeader?> localHeaderAtHeight(Int64 height) async {
      final id = await core.consensus.localChain.blockIdAtHeight(height);
      if (id != null) {
        return await core.dataStores.headers.getOrRaise(id);
      }
      return null;
    }

    final chainSelectionResult = await core.consensus.chainSelection.compare(
      localHeader,
      target,
      commonAncestor,
      localHeaderAtHeight,
      interface.fetchHeaderAtHeight,
    );
    if (chainSelectionResult.isY) {
      await updateTarget(commonAncestor, target, peerId);
    }
  }

  Future<void> remoteCompare(
      BlockHeader target, PeerId peerId, SharedSyncState state) async {
    final clients = clientsF();
    final interface = clients[peerId];
    if (interface == null) {
      throw StateError("Peer not found");
    }
    final localInterface = SortedPeerInterface(clients.values);
    final commonAncestor = await interface.commonAncestor();
    final commonAncestorHeader = await interface.fetchHeader(commonAncestor);

    if (commonAncestorHeader == null) {
      throw StateError("Common ancestor not found");
    }

    final chainSelectionResult = await core.consensus.chainSelection.compare(
      state.target,
      target,
      commonAncestorHeader,
      localInterface.fetchHeaderAtHeight,
      interface.fetchHeaderAtHeight,
    );
    if (chainSelectionResult.isY) {
      await updateTarget(commonAncestorHeader, target, peerId);
    }
  }

  Future<void> updateTarget(
      BlockHeader commonAncestor, BlockHeader target, PeerId provider) async {
    if (_state != null) {
      final state = _state!;
      if (state.target.id == target.id) {
        log.info("Adding provider=${provider.show} to existing sync");
        state.providers.add(provider);
      } else if (state.target.id == target.parentHeaderId) {
        log.info(
            "Retargetting extension of current sync from provider=${provider.show}");
        state.target = target;
        state.providers
          ..clear()
          ..add(provider);
      } else {
        log.info("Canceling current sync");
        await state.cancel();
        _state = null;
        sync(commonAncestor, target, provider);
      }
    } else {
      sync(commonAncestor, target, provider);
    }
  }

  void sync(BlockHeader commonAncestor, BlockHeader target, PeerId provider) {
    log.info(
        "Starting sync to target=${target.id.show} provider=${provider.show}");
    final completer = Completer<void>();

    Stream<void> stream() async* {
      Int64 height = commonAncestor.height;
      Int64 batchTarget = height + 500;
      if (batchTarget > target.height) batchTarget = target.height;
      while (batchTarget <= _state!.target.height) {
        yield* syncHeights(height, batchTarget);
        height = batchTarget + 1;
        batchTarget += 500;
        if (batchTarget > _state!.target.height) {
          batchTarget = _state!.target.height;
        }
      }
      yield null;
      _state = null;
    }

    final sub = stream().listen((_) {},
        onDone: completer.complete,
        onError: completer.completeError,
        cancelOnError: true);

    cancel() async {
      await sub.cancel();
      completer.complete();
    }

    _state = SharedSyncState(
        target: target,
        providers: {provider},
        result: completer.future,
        cancel: cancel);
  }

  Future<void> close() async {
    if (_state != null) {
      await _state!.cancel();
      _state = null;
    }
  }

  Stream<void> syncHeights(Int64 min, Int64 max) async* {
    final interfaces = await interfacesForHeight(max);
    final interface = SortedPeerInterface(interfaces);
    // TODO: parallelism
    BlockHeader? previous;

    yield* Stream.fromIterable(
            List.generate((max - min).toInt() + 1, (i) => min + i))
        .parAsyncMap(16, interface.blockIdAtHeight)
        .parAsyncMap(32, (blockId) async {
          if (blockId == null) {
            throw StateError("Block not found");
          }
          final BlockHeader header;
          final h = await core.dataStores.headers.get(blockId);
          if (h != null) {
            header = h;
          } else {
            final remoteHeader = await interface.fetchHeader(blockId);
            if (remoteHeader == null) {
              throw StateError("Remote header not found");
            }
            header = remoteHeader;
          }
          return header;
        })
        .asyncMap((header) async {
          if (previous != null && header.parentHeaderId != previous!.id) {
            throw StateError("Fork detected");
          }
          // TODO: Don't double-check local
          await PeerBlockchainHandler.checkHeader(core, header);
          previous = header;
          return header;
        })
        .parAsyncMap(32,
            (header) => PeerBlockchainHandler.fetchFullBlock(interface, header))
        .asyncMap((fullBlock) async {
          await PeerBlockchainHandler.checkBody(core, fullBlock);
        });
    if (previous != null) {
      core.consensus.localChain.adopt(previous!.id);
    }
    yield null;
  }

  Future<List<PeerBlockchainInterface>> interfacesForHeight(
      Int64 height) async {
    final clients = clientsF();
    final providers = _state!.providers;
    final providerInterfaces = providers.toList().map((p) => clients[p]!);
    final providersInterface = SortedPeerInterface(providerInterfaces);
    final batchTargetId = await providersInterface.blockIdAtHeight(height);
    if (batchTargetId == null) {
      throw StateError("Block not found");
    }
    final nonProviderClients = Map.of(clients)
      ..removeWhere((id, _) => providers.contains(id));
    final altClients = [
      for (final c in nonProviderClients.values)
        (await c.blockIdAtHeight(height)) == batchTargetId ? c : null
    ].whereType<PeerBlockchainInterface>();
    return [...providerInterfaces, ...altClients];
  }

  static final log = Logger("SharedSync");
}

class SharedSyncState {
  BlockHeader target;
  final Set<PeerId> providers;
  final Future<void> result;
  final Future<void> Function() cancel;

  SharedSyncState(
      {required this.target,
      required this.providers,
      required this.result,
      required this.cancel});
}

Stream<O> parAsyncMapImpl<I, O>(
    int parallelism, Stream<I> stream, Future<O> Function(I) f) {
  bool running = false;
  bool done = false;
  final queue = Queue<Future<O>>();
  final controller = StreamController<O>();
  checkAndPush() async {
    if (running) return;
    running = true;
    while (queue.isNotEmpty) {
      final future = queue.removeFirst();
      final o = await future;
      controller.add(o);
    }
    if (done) {
      controller.close();
    }
    running = false;
  }

  final sub = stream.asyncMap((i) async {
    if (queue.length >= parallelism) {
      await queue.first;
    }
    queue.add(f(i));
    checkAndPush();
  }).listen((_) {}, cancelOnError: true);
  controller.onCancel = sub.cancel;
  sub.onError(controller.addError);
  sub.onDone(() {
    done = true;
    checkAndPush();
  });
  return controller.stream;
}

extension ParStreamOps<T> on Stream<T> {
  Stream<O> parAsyncMap<O>(int parallelism, Future<O> Function(T) f) =>
      parAsyncMapImpl(parallelism, this, f);
}
